### PR TITLE
Extra Units, plus Ratios

### PR DESCRIPTION
**Pull request by Dokotela**
Thanks so much for this package! I definitely needed something like this. I'm very slowly working my way through the [Unified Code for Units of Measure](https://ucum.nlm.nih.gov/), which has all measurements ever (not actually but it's a lot), but your package has almost all of them that I actually need.

I added a couple extra units to your package sub-grams and sub-liter units that I need, plus I added Quarts (because the US insists on using outdated units that make no sense).

The ratio commit I wanted to ask about because I was curious if you had thought about an approach. So many of the units I use are ratios and I was trying to figure out a standard way to work with them. For right now I've done it using a list of enums that get converted into units, and I thought I'd go ahead and see what your thoughts on it were.

### CODE REVIEW

File: lib/models/ratio.dart [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+        return VOLUME.deciliters;
+
+      case Ratio.microGramsPerMilliLiter:
+        return VOLUME.milliliters;
+
+      case Ratio.milliGramsPerMilliliter:
+        return VOLUME.milliliters;
+    }
+  }
+}
```
Comment (ferraridamiano) [inline lib/models/ratio.dart:489]: What's the point with Ratio? You defined it in the model folder even if it is not a model. Just an enum with 2 extensions. Would you like to create a new property?

File: lib/properties/amount_of_substance.dart [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+enum AMOUNTOFSUBSTANCE {
+  moles,
+  femtomoles,
+  picomoles,
+  nanomoles,
+  micromoles,
+  millimoles,
+}
+
+class AmountOfSubstance extends DoubleProperty<AMOUNTOFSUBSTANCE> {
```
Comment (ferraridamiano) [inline lib/properties/amount_of_substance.dart:16]: Seems ok, but tests are missing

File: lib/properties/amount_of_substance.dart [hunk 3]
```diff
@@ -0,0 +1,72 @@
+import 'package:units_converter/models/conversion_node.dart';
+import 'package:units_converter/models/double_property.dart';
+import 'package:units_converter/models/property.dart';
+import 'package:units_converter/models/unit.dart';
+
+//Available length units
+enum AMOUNTOFSUBSTANCE {
```
Comment (ferraridamiano) [inline lib/properties/amount_of_substance.dart:7]: Prefer SCREAMING_SNAKE_CASE to be more coerent

File: lib/properties/mass.dart [hunk 4]
```diff
@@ -110,6 +120,26 @@ class Mass extends DoubleProperty<MASS> {
...
+              name: MASS.nanograms,
+            ),
+            ConversionNode(
+              coefficientProduct: 1e-6,
+              name: MASS.micrograms,
+            ),
+            ConversionNode(
+              coefficientProduct: 0.1,
+              name: MASS.decigrams,
+            ),
```
Comment (ferraridamiano) [inline lib/properties/mass.dart:142]: Missing tests

File: lib/properties/volume.dart [hunk 5]
```diff
@@ -58,6 +65,13 @@ class Volume extends DoubleProperty<VOLUME> {
             VOLUME.usFluidOunces: 'US fl oz',
             VOLUME.imperialGill: 'Imp. gi.',
             VOLUME.usGill: 'US. liq. gi',
+            VOLUME.usQuarts: 'US. liq. qt',
+            VOLUME.femtoliters: 'fl',
+            VOLUME.picoliters: 'pl',
+            VOLUME.nanoliters: 'nl',
+            VOLUME.microliters: 'ul',
+            VOLUME.deciliters: 'dl',
+            VOLUME.centiliters: 'cl',
```
Comment (ferraridamiano) [inline lib/properties/volume.dart:74]: Missing tests

File: lib/utils/temp.txt [hunk 6]
```diff
@@ -0,0 +1,214 @@
+1	Log10 [Unity]
+1	MillionPerSpecimen [Unity]
+1	PerTotalCount [Unity]
+1	Thousand [Unity]
+1	ThousandRedBloodCells [Unity]
+1	OneHundredThousand [Unity]
+1	Million [Unity]
+1	TenToEighth [Unity]
+1	Percent [General Fraction Unit]
```
Comment (ferraridamiano) [inline lib/utils/temp.txt:9]: What's this file?

**Review by ferraridamiano — COMMENTED** [review]
Hi @Dokotela and thank you for this PR. I left few comments. Tell me what do you think

File: lib/models/ratio.dart (hunk 1 above)
Comment (Dokotela) [reply lib/models/ratio.dart:489]: So for this one I wasn't sure the best way to approach it, or where it should go. 

My thought behind it was that there are lots of measurements that aren't basic or primitive measurements (length, mass, etc). But are ratios or combinations of measurements (g/mL, grams/mole, m/s^2 etc). 

I noted that you already had torque as a class which does combine two different measurements into one class. Obviously you could have infinite number of combining units, but even just using the more common ones there quite a few. Now, most of the ones I created could be grouped into just a few different groups (mass/volume, mass/mass, mass/area, amount of substance/mass).

And if you think this is the cleaner way to do it, I can work on that. But my thought was with so many ratios and combinations out there, if you could define a list of the measurements, it's not hard to break them down into their components, convert all of their components, and then return that value. Which I thought was a simpler way of doing it, especially as you started to get more complicated measurements (like the m/s^2).

File: lib/utils/temp.txt (hunk 6 above)
Comment (Dokotela) [reply lib/utils/temp.txt:9]: Apologies, this was a work file I had that I forgot to delete.

Comment (Dokotela) [reply lib/utils/temp.txt]: 6202929c8d20b3ad709e9c1fbfefb7d2cf01b11d

File: lib/properties/amount_of_substance.dart (hunk 3 above)
Comment (Dokotela) [reply lib/properties/amount_of_substance.dart]: 5a36c231ebe5eb099923287b7c01506607d7e638

File: lib/properties/mass.dart (hunk 4 above)
Comment (Dokotela) [reply lib/properties/mass.dart:142]: dda5633b1300292b2d7edc3b4cf36279602d4f99

File: lib/properties/volume.dart (hunk 5 above)
Comment (Dokotela) [reply lib/properties/volume.dart:74]: dda5633b1300292b2d7edc3b4cf36279602d4f99

File: lib/properties/amount_of_substance.dart (hunk 2 above)
Comment (Dokotela) [reply lib/properties/amount_of_substance.dart]: dda5633b1300292b2d7edc3b4cf36279602d4f99

File: lib/models/ratio.dart (hunk 1 above)
Comment (ferraridamiano) [reply lib/models/ratio.dart]: Seems reasonable, give me time to think about this. I'm sorry but I have just few hours in the weekends to dedicate to side project. The rest look good to me!

**Comment by ferraridamiano** [comment]
Ooops, it seems that it's missing tests also in the `getters_test.dart` file

**Comment by ferraridamiano** [comment]
I tought about Ratio. I think that the best way would be to use enhached enums and create a property for each property. For example for density units I would do something like this:
```dart

enum DENSITY {
  gramsPerLiter(MASS.grams, VOLUME.liters),
  gramsPerCubicCentimeter(MASS.grams, VOLUME.cubicCentimeters),
  kilogramsPerLiter(MASS.kilograms, VOLUME.liters),
  kilogramsPerCubicMeter(MASS.kilograms, VOLUME.cubicMeters);

  final MASS numerator;
  final VOLUME denominator;
  const DENSITY(this.numerator, this.denominator);
}

class Density extends DoubleProperty<DENSITY> {
  ///Class for density conversions, e.g. if you want to convert 1 gram per liter
  ///in kilograms per liter:
  ///```dart
  ///var density = Density(removeTrailingZeros: false);
  ///density.convert(Unit(DENSITY.gramsPerLiter, value: 1));
  ///print(DENSITY.kilogramsPerLiter);
  /// ```
  Density(
      {super.significantFigures,
      super.removeTrailingZeros,
      super.useScientificNotation,
      name})
      : super(
          name: name ?? PROPERTY.density,
          mapSymbols: {
            DENSITY.gramsPerLiter: 'g/l',
            DENSITY.gramsPerCubicCentimeter: 'g/cm³',
            DENSITY.kilogramsPerLiter: 'kg/l',
            DENSITY.kilogramsPerCubicMeter: 'kg/m³',
          },
          conversionTree: //...
        );
}
```
What do you think?

Of course we could define a conversionTree similar to other units (and so this example would be very similar ro the final implementation), but as you suggested we could exploit the current conversions of basic properties by decomponing the coposed unit.

I will try in the following weeks to do a preparation work for composed units. In the meantime, you could remove your implementation of Ratio so that we can merge this PR. Then, you could try to add composed units in a separate PR. I don't like very big PR that add to much features

**Comment by Dokotela** [comment]
319db819eafa9a7fac6778c2361d24991200bfb9 I removed ratios, and added in the getters_test that I was missing. I'll take a look and think more about ratios this week.

**Comment by ferraridamiano** [comment]
Perfect! We are almost ready to merge!
Sorry to bother you again, but could you please change the "u" with the "µ" for micrograms and microliters? And could you please write liter subunits related to the liter and not to cubic meters? Thank you!

**Comment by ferraridamiano** [comment]
Thank you very much!

**Comment by ferraridamiano** [comment]
HI @Dokotela, I just added the support for compound units (in the form of a ratio)! Take a look to the [Density](https://github.com/ferraridamiano/units_converter/blob/main/lib/properties/density.dart) property that I just added (just a cuple of units). Now you can add more density units and other conversion. Thank you!

**Pull request merged**